cirq-core/cirq/transformers/routing/initial_mapper.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Dict
+import abc
+
+from cirq import value
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class AbstractInitialMapper(metaclass=abc.ABCMeta):
+    """Base class for creating custom initial mapping strategies.
+
+    An initial mapping strategy is a placement strategy that places logical qubit variables in an
+    input circuit onto physical qubits that correspond to a specified device. This placment can be
+    thought of as a mapping k -> m[k] where k is a logical qubit and m[k] is the physical qubit it
+    is mapped to. Any initial mapping strategy must satisfy two constraints:
+        1. all logical qubits must be placed on the device if the number of logical qubits is <=
+            than the number of physical qubits.
+        2. if two logical qubits interact (i.e. there exists a 2-qubit operation on them) at any
+            point in the input circuit, then they must lie in the same connected components of the
+            device graph induced on the physical qubits in the initial mapping.
+
+    """
+
+    @abc.abstractmethod
+    def initial_mapping(self, circuit: 'cirq.AbstractCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps the logical qubits of a circuit onto physical qubits on a device.
+
+        Args:
+            circuit: the input circuit with logical qubits.
+
+        Returns:
+          qubit_map: the initial mapping of logical qubits to physical qubits.
+        """
+
+
+@value.value_equality
+class HardCodedInitialMapper(AbstractInitialMapper):
+    """Initial Mapper class takes a hard-coded mapping and returns it."""
+
+    def __init__(self, _map: Dict['cirq.Qid', 'cirq.Qid']) -> None:
+        self._map = _map
+
+    def initial_mapping(self, circuit: 'cirq.AbstractCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Returns the hard-coded initial mapping.
+
+        Args:
+            circuit: the input circuit with logical qubits.
+
+        Returns:
+            the hard-codded initial mapping.
+        """
+        return self._map
```

Comment:
What if `circuit.all_qubits()` is not a subset of `self._map.keys()` ?

Reply:
Do we want to have such a check here? I thought this was meant for the user to hard-code any mapping they would like to use despite its feasibility.

Reply:
The docstring of `AbstractInitialMapper` says: 
```
    is mapped to. Any initial mapping strategy must satisfy two constraints:
        1. all logical qubits must be placed on the device if the number of logical qubits is <=
            than the number of physical qubits.
        2. if two logical qubits interact (i.e. there exists a 2-qubit operation on them) at any
            point in the input circuit, then they must lie in the same connected components of the
            device graph induced on the physical qubits in the initial mapping.
```

If we don't have a check, then these conditions are not satisfied.

Reply:
Added a check for that.